backend/telemetry_core/src/aggregator/aggregator_set.rs

```diff
@@ -1,36 +1,85 @@
 use std::sync::atomic::{AtomicUsize, Ordering};
-use std::sync::Arc;
+use std::sync::{Arc, Mutex};
 
 #[derive(Clone)]
 pub struct AggregatorSet(Arc<AggregatorSetInner>);
 
 pub struct AggregatorSetInner {
     aggregators: Vec<Aggregator>,
     next_idx: AtomicUsize,
+    metrics: Mutex<Vec<Metrics>>,
 }
 
 impl AggregatorSet {
     /// Spawn the number of aggregators we're asked to.
     pub async fn spawn(
         num_aggregators: usize,
-        denylist: Vec<String>,
+        opts: AggregatorOpts,
     ) -> anyhow::Result<AggregatorSet> {
         assert_ne!(num_aggregators, 0, "You must have 1 or more aggregator");
 
         let aggregators = futures::future::try_join_all(
-            (0..num_aggregators).map(|_| Aggregator::spawn(denylist.clone())),
+            (0..num_aggregators).map(|_| Aggregator::spawn(opts.clone())),
         )
         .await?;
 
-        Ok(AggregatorSet(Arc::new(AggregatorSetInner {
+        let initial_metrics = (0..num_aggregators).map(|_| Metrics::default()).collect();
+
+        let this = AggregatorSet(Arc::new(AggregatorSetInner {
             aggregators,
             next_idx: AtomicUsize::new(0),
-        })))
+            metrics: Mutex::new(initial_metrics),
+        }));
+
+        // Start asking for metrics:
+        this.spawn_metrics_loops();
+
+        Ok(this)
+    }
+
+    /// Spawn loops which periodically ask for metrics from each internal aggregator.
+    /// Depending on how busy the aggregators are, these metrics won't necessarily be in
+    /// sync with each other.
+    fn spawn_metrics_loops(&self) {
+        let aggregators = self.0.aggregators.clone();
+        for (idx, a) in aggregators.into_iter().enumerate() {
+            let inner = Arc::clone(&self.0);
+            tokio::spawn(async move {
+                loop {
+                    let now = tokio::time::Instant::now();
+                    let metrics = match a.gather_metrics().await {
+                        Ok(metrics) => metrics,
+                        // Any error here is unlikely and probably means that the aggregator
+                        // loop has failed completely.
+                        Err(e) => {
+                            log::error!("Error obtaining metrics (bailing): {}", e);
+                            return;
+                        }
+                    };
+
+                    // Lock, update the stored metrics and drop the lock immediately.
+                    // We discard any error; if something went wrong talking to the inner loop,
+                    // it's probably a fatal error
+                    {
+                        inner.metrics.lock().unwrap()[idx] = metrics;
+                    }
+
+                    // Sleep *at least* 10 seconds. If it takes a while to get metrics back, we'll
+                    // end up waiting longer between requests.
+                    tokio::time::sleep_until(now + tokio::time::Duration::from_secs(10)).await;
```

Comment:
why `sleep_until` instead of just `sleep`?

Reply:
I could have gone either way really, but I went for "aim to have 10 seconds between each update to metrics, unless it takes more than 10 seconds for an update to come through"

Reply:
(To explain the reasoning a bit more; the actual logic to gather metrics is fairly quick, but potentially the message could end up in a queue for a while if we happen to be under heavy load, so if it takes eg 9seconds to get a response, I'd rather put another message into the queue 1 second later than wait another 10 seconds before queueing another message)